yapapi/services/service_runner.py

```diff
@@ -203,6 +223,26 @@ def _change_state(
 
         return instance.state != prev_state
 
+    async def _ensure_alive(self, service: Service):
+        retries_left = self._health_check_retries
+        exc = None
+        while True:
+            if service.is_available and service._ctx:
+                try:
+                    await service._ctx.get_raw_state()
```

Comment:
What `not service._ctx` means? That a service is not up yet?

I would consider adding a method in service that would encapsulate that `get_raw_state` behavior  with a name that descries what it does 

```
            if service.is_available:
                try:
                    await service.is_alive()  # not sure if it is a good name
                    retries_left = self._health_check_retries
                except ApiException as e:
                    ...

# service
async def is_alive() -> None:
    if service._ctx is not None:
        await service._ctx.get_raw_state()
```

Reply:
indeed, worth considering 👍

Reply:
`not service._ctx` -> there's no activity bound to the service

Reply:
okay, @lucekdudek I thought about it and decided against... 

the check on its own is actually not different than just calling `service._ctx.get_raw_state()` and it doesn't give us more information so it doesn't make sense to clutter the API with an additional call like that... 

and if you add retries to `is_alive`, then you need to add more logic on both levels (Service and ServiceRunner) ... and again, if you're operating on the Service level the call to `is_alive` (even with retries inside) doesn't give you much more to work on... I guess it's better to have all logic (retries, the interval and the fact that `_ensure_alive` is a long-running asyncio task) in one place that to spread it all over

Reply:
Everything you do here is on the service level, runner just calls the operations. 
`service.is_available` and `service._ctx.get_raw_state()` are both in service's range of responsibilities.

Why `async def _ensure_alive(self, service: Service):` is not a Service's method in first place. It uses more `service` than `self` suggesting that it would fit better there. 

How about moving it all to service and using it this way?
`                health_check_task = loop.create_task(instance.service.ensure_alive())`

Reply:
My biggest concern is accessing protected fields. It introduces dirt to the code base

Reply:
well, yes, but the concept of "ensure the service is alive" _is_ _not_ a responsibility of a `Service` ... and honestly, given that "protected" fields are more of a convention than a feature in Python, it seems less of a concern than placing logic in places that the logic doesn't belong to...

especially when (when building yapapi), we had agreed to treat underscore-prefixed fields more like a hint to a user that a certain piece is an internal feature and may change without prior warning rather than a true, protected field that no other class should have access to

Reply:
@lucekdudek  updated the implementation based on the voice discussion